Document new context API and fix some docs mishaps (#253)
Roblox/roact: docs/advanced/context.md

!!! danger "Unreleased API"
	This API is not yet available in a stable Roact release.

	It may be available from a recent pre-release or Roact's master branch.

[TOC]

> Roact supports a feature known as context that enables pass values down the tree without having to pass them through props. Roact's Context API is based on [React's Context API](https://reactjs.org/docs/context.html).

Reviewer: This is a nitpick, but wording is a bit odd.
```suggestion
Roact supports a feature called context which enables passing values down the tree without having to pass them through props. Roact's Context API is based on [React's Context API](https://reactjs.org/docs/context.html).
```

```lua
local function ThemedButton(props)
	return Roact.createElement(ThemeContext.Consumer, {
		render = function(theme)
			return Roact.createElement("TextButton", {
				Size = UDim2.new(0, 100, 0, 100),
				Text = "Click Me!",
				TextColor3 = theme.foreground,
				BackgroundColor3 = theme.background,
			})
		end
	})
end
```

The `Provider` component accepts a `value` prop as well as children. Any of its descendants will have access to the value provided to it by using the `Consumer` component like above.

Whenever the `Provider` receives a new `value` prop in an update, any attached `Consumer` components will re-render with the new value. This value could be externally controlled, or could be controlled by state in a component wrapping `Provider`:

```lua
local ThemeController = Roact.Component:extend("ThemeController")

function ThemeController:init()
	self:setState({
		theme = {
			foreground = Color3.new(1, 1, 1),
			background = Color3.new(0, 0, 0),
		}
	})
end

function ThemeController:render()
	return Roact.createElement(ThemeContext.Provider, {
		value = self.state.theme,
	}, self.props[Roact.Children])
end
```

> ## Legacy Context

Reviewer: Are we worried folks will be more willing to add new uses of this now that it's documented in some detail? Calling something "deprecated" hasn't often stopped people before o.O

Author: Maybe we should add more `!!! danger` blocks!

Roact also has a deprecated version of context that pre-dates the stable context API. It will be removed in a future release, but is currently maintained for backwards-compatibility.

Legacy context values cannot be updated dynamically. It is up to the context user to create their own mechanism for updates, probably using a wrapper component and `setState`.

To use it, add new entries to `self._context` in `Component:init()` to create a provider:

```lua
local Provider = Roact.Component:extend("FooProvider")

-- Using a unique non-string key is recommended to avoid collisions.
local FooKey = {}

function Provider:init()
	self._context[FooKey] = {
		value = 5,
	}
end
```

...and read from that same value in `Component:init()` in your consumer component:

```lua
local Consumer = Roact.Component:extend("FooConsumer")

function Consumer:init()
	self.foo = self._context[FooKey]
end

function Consumer:render()
	return Roact.createElement("TextLabel", {

> 		Text = "Foo: " .. self.foo.value,

Reviewer: Should maybe reiterate that this will not update immediately when context updates, but will be up-to-date on the next render?
